pyogrio/_io.pyx

```diff
@@ -2214,3 +2216,190 @@ def ogr_write(
+    object layer_kwargs,
+    bint append=False,
+    dataset_metadata=None,
+    layer_metadata=None,
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    cdef OGRDataSourceH ogr_dataset = NULL
+    cdef OGRLayerH ogr_layer = NULL
+    cdef char **options = NULL
+
+    if not encoding:
+        encoding = locale.getpreferredencoding()
+
+    layer_created = create_ogr_dataset_layer(
+        path, layer, driver, crs, geometry_type, encoding,
+        dataset_kwargs, layer_kwargs, append,
+        dataset_metadata, layer_metadata,
+        &ogr_dataset, &ogr_layer,
+    )
+
+    if geometry_name:
+        opts = {"GEOMETRY_NAME": geometry_name}
+    else:
+        opts = {}
+    options = dict_to_options(opts)
+
+    stream_capsule = arrow_obj.__arrow_c_stream__()
+    write_arrow_stream_capsule(ogr_layer, stream_capsule, geometry_name, options)
+
+    ### Final cleanup
+    if options != NULL:
+        CSLDestroy(options)
+        options = NULL
+
+    if ogr_dataset != NULL:
+        GDALClose(ogr_dataset)
+
+        # GDAL will set an error if there was an error writing the data source
+        # on close
+        exc = exc_check()
+        if exc:
+            raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+
+IF CTE_GDAL_VERSION >= (3, 8, 0):
+
+    cdef write_arrow_stream_capsule(OGRLayerH destLayer, object capsule, str geometry_name, char** options = NULL):
+        cdef ArrowSchema schema
+        cdef ArrowArray array
+
+        cdef ArrowArrayStream* stream = <ArrowArrayStream*>PyCapsule_GetPointer(
+            capsule, "arrow_array_stream"
+        )
+        if stream == NULL:
+            raise RuntimeError("Could not extract valid Arrow array stream.")
+
+        if stream.release == NULL:
+            raise RuntimeError("Arrow array stream was already released.")
+
+        if stream.get_schema(stream, &schema) != 0:
+            stream.release(stream)
+            raise RuntimeError("Could not get Arrow schema from stream.")
+
+        try:
+            create_fields_from_arrow_schema(destLayer, &schema, options, geometry_name)
+        except Exception as e:
+            schema.release(&schema)
+            stream.release(stream)
+            raise e
+
+        while True:
+            if stream.get_next(stream, &array) != 0:
+                schema.release(&schema)
+                stream.release(stream)
+                raise RuntimeError("Error while accessing batch from stream.")
+
+            # We've reached the end of the stream
+            if array.release == NULL:
+                break
+
+            if not OGR_L_WriteArrowBatch(destLayer, &schema, &array, options):
+                if array.release != NULL:
+                    array.release(&array)
+
+                schema.release(&schema)
+                stream.release(stream)
+
+                exc = exc_check()
+                if exc:
+                    raise exc
```

Comment:
I think we actually typically catch the `CPLE_..` error and reraise that with one of our generic error classes?

For example I see this pattern:

```
            except CPLE_BaseError as exc:
                raise FeatureError(str(exc))
```

But then what error class to use? (`FeatureError` or `FieldError`?)

Reply:
Probably `DataSourceError` since it is for the batch of features rather than individual features.

Reply:
Used DataLayerError following our documentation, as it involves an error with a specific layer ("Errors relating to working with a single OGRLayer") and not with the whole dataset ("Errors relating to opening or closing an OGRDataSource (with >= 1 layers)")